File: DP/g.py

tests =  int(input())
for _ in range(tests):
    n_weights = int(input())
    weights = []
    values = []
    result = 0
    for _ in range(n_weights):
        value , weight = map(int, input().split())
        weights.append(weight)
        values.append(value)
    n = len(values)
    people = int(input())
    max_weights = []
    for __ in range(people):
        max_weights.append(int(input()))
    dp = [[0 for _ in range(max(max_weights) + 1)] for _ in range(n + 1)]

    for weight in max_weights:
        max_weight = weight
        for i in range(1, n + 1):
            for w in range(max_weight + 1):
                if weights[i-1] <= w:
                    dp[i][w] = max(dp[i-1][w], dp[i-1][w-weights[i-1]] + values[i-1])
                else:
                    dp[i][w] = dp[i-1][w]

        person_max = dp[n][max_weight]
        result += person_max
    print(result)

File: 12.04/g.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int tests;
    cin >> tests;

    while (tests--) {
        int n_weights;
        cin >> n_weights;

        vector<int> weights(n_weights);
        vector<int> values(n_weights);
        
        for (int i = 0; i < n_weights; ++i) {
            cin >> values[i] >> weights[i];
        }

        int n = values.size();

        int people;
        cin >> people;

        vector<int> max_weights(people);

        for (int i = 0; i < people; ++i) {
            cin >> max_weights[i];
        }

        int max_max_weight = *max_element(max_weights.begin(), max_weights.end());

        vector<vector<int>> dp(n + 1, vector<int>(max_max_weight + 1, 0));

        for (int i = 1; i <= n; ++i) {
            for (int w = 0; w <= max_max_weight; ++w) {
                if (weights[i - 1] <= w) {
                    dp[i][w] = max(dp[i - 1][w], dp[i - 1][w - weights[i - 1]] + values[i - 1]);
                } else {
                    dp[i][w] = dp[i - 1][w];
                }
            }
        }

        int result = 0;
        for (int weight : max_weights) {
            result += dp[n][weight];
        }

        cout << result << endl;
    }

    return 0;
}
